### ISSUE TITLE
Deprecate packageOverrides?

### ISSUE THREAD

**Issue by lheckemann** [NixOS/nixpkgs#43266]
`packageOverrides` is replaced by overlays as far as functionality goes, and has a number of disadvantages:

- Learning/documentation overhead. We have 3 distinct things with the word override in them: `packageOverrides`, `pkg.override` introduced by `callPackage`, and `pkg.overrideAttrs` introduced by `stdenv.mkDerivation`. This is very confusing to beginners (it was to me as well, and we didn't have overlays at the time). Additionally, `packageOverrides` shares a space with overlays while being less powerful.
- Unwieldy composition. I seem to recall there being some trick involving something like `self: let super = self.pkgs;` to get an approximation of what overlays provide, but it inrtoduces a fixpoint in each layer AFAIU. I personally don't know exactly how this works, but overlays are pretty clear in this respect.

For this reason, I think it would be nice to deprecate `packageOverrides` (remove it from docs in favour of overlays, add a warning when nixpkgs is evaluated with `config.packageOverrides` set) and eventually remove it completely, leaving perhaps only an explanatory error message. The warning/error messages should explain how to turn a `packageOverrides` function into an overlay to ease use of existing third-party documentation.

Any opinions on whether this is a good idea at all, further practical considerations, or any other relevant commentary welcome!

**Comment by grahamc** [comment]
cc @nbp

**Comment by nbp** [comment]
If I do not make any mistake this should almost be a backward compatible overlay for adding the feature back:

```nix
self: super:

(super.config.packageOverrides or (super: {})) super
```

The remaining issue is about modifying any bootstrap packages. This should be fixed by including all the stages of the bootstrap as attribute names within the main fix-point.

**Comment by lheckemann** [comment]
Given [the current implementation of packageOverrides](https://github.com/NixOS/nixpkgs/blob/d3ed77f1fb1d4e7ef7ec7f63844c9257bc243be8/pkgs/top-level/stage.nix#L109-L118), yes, I think that should work :D

**Comment by lheckemann** [comment]
Opened PR #43560 to implement this.

**Comment by jcrben** [comment]
Overlays seem tricky to use correctly. Therefore I think deprecation at this point could on net cause more UX damage than non-deprecation. 🤔  

Right now, they aren't showing up for me on my mac in `nix-env -q` altho I can install them OK (might be better on NixOS machine). Looks like https://github.com/NixOS/nixpkgs/issues/26487#issuecomment-365951435 ran into similar problems altho I don't understand the fix. And https://github.com/NixOS/nixpkgs/issues/25264 has a long discussion about how to get overlays to be picked up by tools, but I don't see any sign of commits to fix the situation. Maybe hardening overlays and triaging / resolving the open issues on them (e.g., https://github.com/NixOS/nixpkgs/issues/24907 https://github.com/NixOS/nixpkgs/issues/28558 ) before deprecating.

And the existing docs on overlays could be better as there are nuances that could to be called out a bit more strongly, e.g. single file needs to return a list while overlays in a directory are functions.

**Comment by ckauhaus** [comment]
I've converted an overlay to a packageOverrides construct lately because the overlay took too much memory to evaluate. It seems that a complete pkgs attrset mut be constructed on each fixpoint iteration. packageOverrides are more compact on memory usage exactly because they involve a local fixpoint.

This is by no way meant to stop overlays, but it should be taken into consideration before deprecating packageOverrides too early.

**Comment by lheckemann** [comment]
packageOverrides is implemented as a stage of nixpkgs which behaves just like an overlay, so I'm not sure I understand how this makes a difference.

Although at this point it almost seems to me like removing overlays again might be better :see_no_evil:

**Comment by ckauhaus** [comment]
I got random GC errors like "out of memory" or "too many heap sections" while using overlays (something similar to #40456). After converting overlays to packageOverrides, the errors disappeared. Actually no idea if this was just luck or a real solution.

**Comment by asymmetric** [comment]
I currently use `packageOverrides` to declare a set of packages in my `~/.config/nixpkgs/config.nix` (explained [here](https://nixos.org/nixpkgs/manual/#sec-declarative-package-management)).

How would I do that - declaratively - with overlays? i.e. I don't want to have to run `nix-env` for each package in the overlay, but rather have a "meta-package" of sorts that bundles all the packages I want.

**Comment by lheckemann** [comment]
@asymmetric As documented in the warning in my PR https://github.com/NixOS/nixpkgs/pull/43560/files#diff-69df24a05ec3d577cb638113e9837ae1R119 :
take only the value of packageOverrides, and replace `pkgs: with pkgs;` with `self: super: with super;`, then put it in a .nix file in `~/.config/nixpkgs/overlays`.

**Comment by lheckemann** [comment]
As mentioned on the issue, will do an RFC eventually

**Comment by ShamrockLee** [comment]
Is this still revalent?

**Comment by lheckemann** [comment]
Yes

**Comment by asymmetric** [comment]
@lheckemann if this is still relevant (which I think it is), then maybe it should be turned into an RFC. I don't think @edolstra's mild opposition to this should stop you from writing it, for the record :)

**Comment by lheckemann** [comment]
>As mentioned on the issue, will do an RFC eventually

for some value of eventually…

**Comment by ThibautMarty** [comment]
In the meantime, we may change the documentation without deprecating `packageOverrides`, what do you think?

I was helping a newcomer that used `pkgs.buildEnv` to install a set of packages at once. The method [documented here](https://github.com/NixOS/nixpkgs/blob/b6fa4975cf627604fd2f7ea7b1987f665f99a771/doc/using/configuration.chapter.md#L12) uses `packageOverrides` and not overlays. His package set didn't contain a package that he patched in an overlay but the the original one which was confusing for him.
The fact that a newcomer seems to be able to create an overlay but is guided to use `packageOverrides` seems wrong to me.

I guess the use of `packageOverrides` here (and in other parts of the documentation) is for historical reasons only, so I would be in favor of replacing it with overlays.
I can make a patch if no one objects..

**Comment by YoshiRulz** [comment]
So to clarify, should this `config.nix` section
```nix
packageOverrides = pkgs: {
	myPackages = pkgs.buildEnv {
		name = "my-packages";
		paths = [ pkgs.hello ];
	};
};
```
be replaced with
```nix
final: prev: {
	myPackages = prev.buildEnv {
		name = "my-packages";
		paths = [ prev.hello ];
	};
}
```
as an overlay `my-packages.nix`? It seems to work at least.

**Comment by lheckemann** [comment]
@YoshiRulz yes, that looks mostly correct (though usually you'd want to get `hello`, being a package, from `final` so that it can still be affected by later overlays).
